### PR TITLE
Returning list of op names in get_graph_ops and extended error atoms to all TF error codes

### PR DESCRIPTION
**Pull request by anshuman23**
So this PR covers the changes you had last requested @josevalim. Other than that I have just removed some unused code from before. Also ensured I am using `enif_make_binnary` wherever needed instead of `enif_make_string`:
- __Extending the error coverage__: The C function `error_to_string` handles that now and an example of a faulty graph reading attempt would look like this:
```elixir
iex(1)> graph = Tensorflex.read_graph "Makefile"
{:error, :invalid_argument}
```
- __Returning list of all operations in graph instead of directly printing them__: This has also been taken care of, and now the function is called `get_graph_ops` which returns a List of strings (op names in the graph). An example is as follows:
```elixir
iex(1)> graph = Tensorflex.read_graph("classify_image_graph_def.pb")
2018-05-17 23:36:16.488469: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that    this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA 2018-05-17 23:36:16.774442: W             tensorflow/core/framework/op_def_util.cc:334] OpBatchNormWithGlobalNormalization is deprecated. It will cease to work in   GraphDef version 9. Use tf.nn.batch_normalization().
Successfully imported graph
#Reference<0.1610607974.1988231169.250293>

iex(2)> op_list = Tensorflex.get_graph_ops graph
["softmax/biases", "softmax/weights", "pool_3/_reshape/shape",
 "mixed_10/join/concat_dim", "mixed_10/tower_2/conv/batchnorm/moving_variance",
 "mixed_10/tower_2/conv/batchnorm/moving_mean",
 "mixed_10/tower_2/conv/batchnorm/gamma",
 "mixed_10/tower_2/conv/batchnorm/beta", "mixed_10/tower_2/conv/conv2d_params",
 "mixed_10/tower_1/mixed/conv_1/batchnorm/moving_variance",
 "mixed_10/tower_1/mixed/conv_1/batchnorm/moving_mean",
 "mixed_10/tower_1/mixed/conv_1/batchnorm/gamma",
 "mixed_10/tower_1/mixed/conv_1/batchnorm/beta",
 "mixed_10/tower_1/mixed/conv_1/conv2d_params",
 "mixed_10/tower_1/mixed/conv/batchnorm/moving_variance",
 "mixed_10/tower_1/mixed/conv/batchnorm/moving_mean",
 "mixed_10/tower_1/mixed/conv/batchnorm/gamma",
 "mixed_10/tower_1/mixed/conv/batchnorm/beta",
 "mixed_10/tower_1/mixed/conv/conv2d_params",
 "mixed_10/tower_1/conv_1/batchnorm/moving_variance",
 "mixed_10/tower_1/conv_1/batchnorm/moving_mean",
 "mixed_10/tower_1/conv_1/batchnorm/gamma",
 "mixed_10/tower_1/conv_1/batchnorm/beta",
 "mixed_10/tower_1/conv_1/conv2d_params",
 "mixed_10/tower_1/conv/batchnorm/moving_variance",
 "mixed_10/tower_1/conv/batchnorm/moving_mean",
 "mixed_10/tower_1/conv/batchnorm/gamma",
 "mixed_10/tower_1/conv/batchnorm/beta", "mixed_10/tower_1/conv/conv2d_params",
 "mixed_10/tower/mixed/conv_1/batchnorm/moving_variance",
 "mixed_10/tower/mixed/conv_1/batchnorm/moving_mean",
 "mixed_10/tower/mixed/conv_1/batchnorm/gamma",
 "mixed_10/tower/mixed/conv_1/batchnorm/beta",
 "mixed_10/tower/mixed/conv_1/conv2d_params",
 "mixed_10/tower/mixed/conv/batchnorm/moving_variance",
 "mixed_10/tower/mixed/conv/batchnorm/moving_mean",
 "mixed_10/tower/mixed/conv/batchnorm/gamma",
 "mixed_10/tower/mixed/conv/batchnorm/beta",
 "mixed_10/tower/mixed/conv/conv2d_params",
 "mixed_10/tower/conv/batchnorm/moving_variance",
 "mixed_10/tower/conv/batchnorm/moving_mean",
 "mixed_10/tower/conv/batchnorm/gamma", "mixed_10/tower/conv/batchnorm/beta",
 "mixed_10/tower/conv/conv2d_params", "mixed_10/conv/batchnorm/moving_variance",
 "mixed_10/conv/batchnorm/moving_mean", "mixed_10/conv/batchnorm/gamma",
 "mixed_10/conv/batchnorm/beta", "mixed_10/conv/conv2d_params",
 "mixed_9/join/concat_dim", ...]

```

### CODE REVIEW

File: c_src/Tensorflex.c [hunk 1]
```diff
@@ -119,6 +118,48 @@ static ERL_NIF_TERM new_op(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
...
+      break;
+    case TF_UNIMPLEMENTED: strcpy(error,"unimplemented");
+      break;
+    case TF_INTERNAL: strcpy(error,"internal");
+      break;
+    case TF_UNAVAILABLE: strcpy(error,"unavailable");
+      break;
+    case TF_DATA_LOSS: strcpy(error,"data_loss");
+      break;
+    default: strcpy(error,"unlisted_code");
```
Comment (josevalim) [inline c_src/Tensorflex.c]: Maybe we could have each of those call `enif_make_atom` so we don't have to allocate the string in the first place, just to convert it to an atom?

File: c_src/Tensorflex.c [hunk 2]
```diff
@@ -162,6 +206,39 @@ static ERL_NIF_TERM read_graph(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv
...
+  ERL_NIF_TERM op_list_eterm;
+  TF_Operation *op_temp;
+  ErlNifBinary erl_str;
+  op_list = malloc(sizeof(ERL_NIF_TERM)*n_ops);
+  pos = 0;
+  
+  char op_name[BASE_STRING_LENGTH];
+  for(int i=0; i<n_ops; i++) {
+    op_temp = TF_GraphNextOperation(*graph, &pos);
+    strcpy(op_name, (char*) TF_OperationName(op_temp));
```
Comment (josevalim) [inline c_src/Tensorflex.c]: Do we need to copy the string here? Given that we are already copying it later to the binary with `memcpy`?

**Comment by josevalim** [comment]
Comments added. Other than that, it looks great to go! :+1:

What next steps do you have in mind? :)

File: c_src/Tensorflex.c (hunk 1 above)
Comment (anshuman23) [reply c_src/Tensorflex.c]: Fixed this in the latest commit. I was needlessly allocating a string for this. Renamed the function to `error_to_atom` now.

File: c_src/Tensorflex.c (hunk 2 above)
Comment (anshuman23) [reply c_src/Tensorflex.c]: Again, this wasn't needed and I have fixed this in the latest commit. I apologize for the poorly written code here.

**Comment by anshuman23** [comment]
Thanks for the feedback @josevalim!

I have been thinking about the next steps, and the logical progression of adding functions should eventually get to the point where we can run the loaded graph against our own inputs and generate prediction outputs. For this to work, a lot of functions need to be added first: particularly ones that will allow us to read Tensors supplied by the user. I am thinking over how I would accomplish this and then I will submit a PR for review. These TF_Tensor functions are very important and they would have to encompass a wide variety of inputs. 

I might also keep importing graph based functions from Tensorflow as and when required (like the operation list function `get_graph_ops`) After that, I think some more graph based functions will be needed that will use both the tensor functionality and the graph functions. Finally, we will need to add support for creating and running Sessions.

**Comment by josevalim** [comment]
@anshuman23 agreed! :heart:

Let's go with this plan and continue exploring the API and the "user story" but note that at some point we will have to "sit down" and write documentation and tests.

Also, can you please send an e-mail to the beamcommunity mailing list with your progress so far and your plans for the next week? What is coming next can be a very quick digest, like this one that you just sent!

**Comment by anshuman23** [comment]
Thanks @josevalim, and I completely agree on taking time out to write documentation and tests. I will incorporate that into my plans as well. :)

Yes, I'll send that email right away. Thanks again for all the help!

**Pull request merged**